Docs 7126 scatter gather v4.2 (#1169)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/scatter-gather-concept.adoc

include::_attributes.adoc[]
endif::[]

> 

Reviewer: Instructions for edit:

Link to the PR: https://github.com/mulesoft/docs-mule-runtime/pull/1169
Type of edit requested : Standard
Due date: Friday 02/07
Additional details: There are some instances were I used passive voice or made other mistakes because I didn't know how to explain that scenario, so I need your recommendations to address whatever issues you detect.

Reviewer: Hi Cristian,

I had to read and reread this text several times, and I still can't visualize the relationships among routes, Scatter-Gather component, and flows, or between routes that complete successfully with events but routes that fail with errors, or between the processing flow and the branch created by the SG.

So the editing here is pretty basic, and I have lots of questions. I am hoping that by asking them I give you some idea of what someone new to this might be confused about, but I am afraid I cannot offer much in the way of suggested rewording.

> The Scatter-Gather component is a routing event processor that processes a Mule event through different parallel processing routes that contain one or more event processors. Each route receives a reference to the Mule event and executes a sequence of one or more event processors. Each of these routes uses a separate thread and creates a new Mule event with its own payload, attributes, and variables. The Scatter-Gather component then combines the Mule events returned by each processing route into a new Mule event that is passed to the next event processor only after every route completes successfully.

Reviewer 2: the new event is not generated by the routes, but by the processors in the route. If a route is empty or has read-only (void) operations, the event at the end of the route is the same as in the beginning.

> The Scatter-Gather component is a routing event processor that processes a Mule event through different parallel processing routes that contain one or more event processors. Each route receives a reference to the Mule event and executes a sequence of one or more event processors. Each of these routes uses a separate thread to execute the event processors, and the resulting Mule event can be either the same Mule event without modifications, or a new Mule event with its own payload, attributes, and variables. The Scatter-Gather component then combines the Mule events returned by each processing route into a new Mule event that is passed to the next event processor only after every route completes successfully.

Reviewer: ```suggestion
The Scatter-Gather component is a routing event processor that processes a Mule event through different parallel processing routes that contain one or more event processors. Each route receives a reference to the Mule event and executes a sequence of one or more event processors. Each of these routes uses a separate thread to execute the event processors, and the resulting Mule event can be either the same Mule event without modifications or a new Mule event with its own payload, attributes, and variables. The Scatter-Gather component then combines the Mule events returned by each processing route into a new Mule event that is passed to the next event processor only after every route completes successfully.
```

Reviewer: You might want to make the previous description less detailed, if you are going to provide details following the diagram.

Reviewer: Maybe...:
```suggestion
The Scatter-Gather component is a routing event processor that processes a Mule event through different parallel processing routes that contain one or more event processors. Each route receives a reference to the Mule event and executes a sequence of one or more event processors. Each of these routes uses a separate thread to execute the event processors, and the resulting Mule event can be either the same Mule event or a new Mule event. The Scatter-Gather component then combines the Mule events returned by each processing route into a new Mule event that is passed to the next event processor only after every route completes successfully.
```

Reviewer: Also, I looked at the previous text and found the explanations of the source of the terms "scatter" and "gather" helpful, although I do not understand why it is necessary to have parallel processing routes: does each route process the one event in different ways, to try to find the one solution?

The Scatter-Gather component executes each route in parallel, not sequentially. Parallel execution of routes can greatly increase the efficiency of your Mule application and may provide more information than sequential processing.

The following diagram details the behavior of the Scatter-Gather component:

image::mruntime-scatter-gather.png[Diagram of Scatter-Gather Component]

> <1> The Scatter-Gather component receives a Mule event and sends a reference of this Mule event to each processing route.

Reviewer: Why are there more than one processing route? Does each involve a different scenario or test or...?

> <2> Each of the processing routes start executing in parallel. Once all processors inside a route finish processing, the route generates a new Mule event as a result.

Reviewer 2: this is not true. refer to the previous comment

> <2> Each of the processing routes start executing in parallel. Once all processors inside a route finish processing, the route returns a Mule event, which can be either the same Mule event without modifications, or a new Mule event created by the processors in the route as a result of the modifications applied.

Reviewer: ```suggestion
<2> Each of the processing routes starts executing in parallel. After all processors inside a route finish processing, the route returns a Mule event, which can be either the same Mule event without modifications or a new Mule event created by the processors in the route as a result of the modifications applied.
```

Reviewer: But...what has this parallel processing achieved, if the result is still an event that needs processing?

> <2> Each of the processing routes start executing in parallel. Once all processors inside a route finish processing, the route returns a Mule event, which can be either the same Mule event without modifications, or a new Mule event created by the processors in the route as a result of the modifications applied.

Reviewer: ```suggestion
<2> Each of the processing routes start executing in parallel. After all processors inside a route finish processing, the route returns a Mule event, which can be either the same Mule event without modifications or a new Mule event created by the processors in the route as a result of the modifications applied.
```

<3> After all processing routes have finished execution, the Scatter-Gather component creates a new Mule event that combines all resulting Mule events from each route, and then passes the new Mule event to the next component in the flow.

> 

Reviewer: As mentioned above, consider a way to deduplicate info between the intro text above the diagram and the steps below. There is a lot of repetition in this topic set overall, between the intro, the text describing the diagram, and the following example scenarios using a Try scope.

> WARNING: A Scatter-Gather component cannot be configured with only one route. A Mule application with this configuration throws an exception and does not start, at least two routes are necessary in a Scatter-Gather configuration to be valid.

Reviewer: ```suggestion
WARNING: Configure your Scatter-Gather component to have at least two routes; otherwise, your Mule application throws an exception and does not start. 
```

== Error Handling Inside Scatter-Gather Routes

> You can use a Try scope in each route of a Scatter-Gather component to handle any errors that might be generated by a route’s event processors. Unhandled errors that are thrown because of a problem in a route cause the Scatter-Gather component to also throw an error of type `MULE:COMPOSITE_ERROR`. The component throws this error only after every route completes, either with a successful execution or with an error that causes the route to fail. Event processing does not proceed further than the Scatter-Gather component in the flow. Instead, the flow branches to your error-handling event processors. The `MULE:COMPOSITE_ERROR` error object gathers together not only the errors from the route that failed, but also the Mule events from completed routes, so that your application can still process Mule events from the routes that completed by using your error-handling event processors.

Reviewer: PLEASE read carefully to determine that I have not changed your intended meaning:
```suggestion
You can use a Try scope in each route of a Scatter-Gather component to handle any errors that might be generated by a route’s event processors. After every route executes, if any has failed with an error, then the Scatter-Gather component throws an error of type `MULE:COMPOSITE_ERROR`, and event processing does not proceed past the Scatter-Gather component in the flow. Instead, the flow branches to your error-handling event processors. 

Because the `MULE:COMPOSITE_ERROR` error object gathers not only errors from routes that failed but also Mule events from successfully completed routes, your application can use the error-handling event processors to process Mule events from the routes that completed.
```

> You can use a Try scope in each route of a Scatter-Gather component to handle any errors that might be generated by a route’s event processors. Unhandled errors that are thrown because of a problem in a route cause the Scatter-Gather component to also throw an error of type `MULE:COMPOSITE_ROUTING`. The component throws this error only after every route completes, either with a successful execution or with an error that causes the route to fail. Event processing does not proceed further than the Scatter-Gather component in the flow. Instead, the flow branches to your error-handling event processors. The `MULE:COMPOSITE_ROUTING` error object gathers together not only the errors from the route that failed, but also the Mule events from completed routes, so that your application can still process Mule events from the routes that completed by using your error-handling event processors.

Reviewer: Do we need to introduce or explain what a Try scope is and how it works?

Reviewer: If a Try scope is beneficial, why would a route not include one?

Reviewer: If the Try scope "handles errors generated by route's event processors," are those still considered "unhandled" errors, or does all the text following the first sentence above simply repeat all the previous text about the SG `MULE:COMPOSITE_ROUTING` error that results and how it is handled?

Could we skip the repetition and use instead:
```suggestion
You can use a Try scope in each route of a Scatter-Gather component to handle any errors that might be generated by a route’s event processors and to avoid causing the Scatter-Gather component to throw an `MULE:COMPOSITE_ROUTING` error. To illustrate how this works, consider the following two cases:
```

> To illustrate how this works, consider the following two cases:

Reviewer: See suggestions for reducing text and moving this sentence above.

> * The routes in a Scatter-Gather component each contain a Try scope. One of the routes generates an error that is successfully handled by that route’s Try scope, so the route is still able to generate a Mule event. The Scatter-Gather component consolidates the Mule events from all routes into a new Mule event, as usual, and passes the consolidated event to the next event processor.

Reviewer 2: only if the error handler within the route is `on-error-continue`. For `on-error-propagate` an error will be generated for the route still.

> * The routes in a Scatter-Gather component each contain a Try scope. One of the routes generates an error that is successfully handled by that route’s Try scope through an `on-error-continue`, so the route is still able to generate a Mule event. The Scatter-Gather component consolidates the Mule events from all routes into a new Mule event and passes the consolidated event to the next event processor.

Reviewer: This and the following suggestion attempt to make the two cases more distinct and parallel>
It might be a good idea to illustrate each case with a diagram as well, because this seems pretty complicated.
```suggestion
* The routes in a Scatter-Gather component each contain a Try scope. 
One of the routes generates an error that is successfully handled by that route’s Try scope through an `on-error-continue` operation, so the route is completed successfully. The Scatter-Gather component does not throw an error; it consolidates the Mule events from all routes into a new Mule event and passes the consolidated event to the next event processor.
```

> * An error occurs in a route. This route either does not contain a Try scope or the error is of a type that the Try scope cannot handle. Although all the other routes complete processing, the failed route causes the Scatter-Gather component to throw a `MULE.COMPOSITE ERROR` error, which the flow passes to the Error Handler component that you have configured for your flow. The error-handling event processors are able to process the Mule events from the completed routes.

Reviewer 2: `MULE.COMPOSITE ERROR` -> `MULE:COMPOSITE_ROUTING`

> * An error occurs in a route. This route does not contain a Try scope, or contains a Try scope with an error handler that cannot handle the type or error, or the error handler is an `on-error-propagate`. Although all the other routes complete processing, the failed route causes the Scatter-Gather component to throw a `MULE:COMPOSITE_ROUTING` error, which the flow passes to the Error Handler component that you have configured for your flow. The error-handling event processors are able to process the Mule events from the completed routes.

Reviewer: ```suggestion
* One of the routes in a Scatter-Gather component does not contain a Try scope or contains a Try scope with an error handler that cannot handle the error type, or the error handler is an `on-error-propagate` type.
An error occurs in this route, causing the route to fail, which in turn causes the Scatter-Gather component to throw a `MULE:COMPOSITE_ROUTING` error. The flow branches to your error-handling event processors, which are able to process the Mule events from the completed routes.
```

Reviewer: I still don't understand how the failed routes are resolved. Maybe that's not necessary?

include::_attributes.adoc[]
endif::[]

The Scatter-Gather component is a routing event processor that processes a Mule event through different parallel processing routes that contain one or more event processors. Each route receives a reference to the Mule event and executes a sequence of one or more event processors. Each of these routes uses a separate thread to execute the event processors, and the resulting Mule event varies according to the following:

> * If the processors in the route apply any modification to the Mule event received, the processors create a new Mule event with its own payload, attributes, and variables. When the route completes, it returns the new Mule event.

Reviewer 2: this is general to all mule chains, i don't think it is necessary to make it explicit here for scatter gather (unless we are clarifying this for all routes, which i don't think is the case)

To illustrate how this works, consider the following two cases:

* The routes in a Scatter-Gather component each contain a Try scope. One of the routes generates an error that is successfully handled by that route’s Try scope through an `on-error-continue`, so the route is still able to generate a Mule event. The Scatter-Gather component consolidates the Mule events from all routes into a new Mule event and passes the consolidated event to the next event processor.

> * An error occurs in a route. This route either does not contain a Try scope or the error is of a type that the Try scope cannot handle. Although all the other routes complete processing, the failed route causes the Scatter-Gather component to throw a `MULE:COMPOSITE_ROUTING` error, which the flow passes to the Error Handler component that you have configured for your flow. The error-handling event processors are able to process the Mule events from the completed routes.

Reviewer 2: `or the error is of a type that the Try scope cannot handle` -> `or the error is of a type that the Try scope doesn't have a handler for, or the handler is an "on-error-propagate"`

`{var1: "var1", var2: "newValue", var3: ["appleVal, bananaVal, otherVal"], var4: "val4"}`

== Handle Timeout Errors in a Scatter-Gather

> You can configure a timeout for a Scatter-Gather component. If a route does not complete processing before the period of time set in the timeout expires, the route throws a `MULE:TIMEOUT` error. This error is then handled the same way as any other error generated from a route: after each route completes (either by completing its last event processor or by throwing a `MULE:TIMEOUT` error), the successful results and errors are collected together and thrown by the Scatter-Gather component as a `MULE:COMPOSITE_ROUTING` error, which can then be processed in an error handler.

Reviewer: ```suggestion
If you configure a timeout for a Scatter-Gather component and a route does not complete processing before the timeout expires, the route throws a `MULE:TIMEOUT` error. This error is then handled the same way as any other error generated from a route: after each route completes (either by processing success or by throwing a `MULE:TIMEOUT` error), the successful results and errors are collected together in the Scatter-Gather component `MULE:COMPOSITE_ROUTING` error, which is then processed in an error handler.
```

> == Example

Reviewer: Do we need both this example scenario and the two cases illustrated above?
